### DISCUSSION TITLE
(Done) April 9th 2025: Discussion: fast-edn and its broader implications

### DISCUSSION THREAD

**Discussion by frenchy64** [madclj/madclj.com#13] [Announcements]
👍 to RSVP

https://madclj.com/

Wednesday, April 9th, 2025
6:30 PM to 9:00 PM CT

[StartingBlock Madison](https://www.google.com/maps/search/?api=1&query=43.081245%2C%20-89.374054)
821 E Washington Ave 2nd floor · Madison, WI

[fast-edn](https://github.com/tonsky/fast-edn) claims to parse EDN 6x faster than Clojure. It is based on [charred](https://github.com/cnuernber/charred), a similar library but for JSON.

Questions:
- How does fast-edn manage to outperform other approaches?
- Are you aware of applications that might run faster with 6x faster EDN parsing?
- Can these insights be used elsewhere in the Clojure compiler?
- fast-edn claims parity with clojure.edn. Which correspondences are important?
- How would you test these correspondences hold?
- Why is clojure.edn separate from Clojure's parser, despite EDN being a subset of Clojure syntax?
- fast-edn is faster than transit. When would use choose transit over fast-edn?

Schedule:

- 6:30pm arrive, food & drink provided
- 7pm, start discussion
- 9pm end meetup

Organizer notes:
- need paper plates + sparkle water

**Comment by frenchy64** [comment]
@madclj/meetup The April 9th meetup will be a overview and discussion about fast-edn and its broader implications.

**Reply by brunchboy** [reply]
This sounds interesting! Unfortunately I’ll be landing in Mexico City around then. ✈️ 😄

**Comment by ericnormand** [comment]
I can bring paper plates and water.

**Reply by frenchy64** [reply]
Thanks Eric!

I need to look into why the RSVP count doesn't show up on madclj.com.

**Comment by sivakusayan** [comment]
Something came up and I can't attend tomorrow 🙁  hope ya'll have fun though!

**Comment by frenchy64** [comment]
I'd like to add a quick group hacking session today if we have time.

The main theme will be improving test coverage of fast-edn's parity against clojure.edn. i.e., writing tests that increase evidence for fast-edn being a drop-in replacement for clojure.edn.

I have a specific idea that I can drive for:

There are already good tests in fast-edn, but I think generative tests can be improved. Currently, it tests something like:

```clojure
(forall [x gen/any] (= (edn/read-string (pr-str x)) (fast/read-string (pr-str x))))
```

This only tests that the output of `pr-str` is parsed the same way in both. This does not test things like whitespace since `pr-str` always prints the same, or invalid edn.

For example, this tests if `"[:a :b]"` is parsed the same but not `"[:a, :b]"` or `"[:a :b])"`.

**Comment by djwhitt** [comment]
Can't make it tonight unfortunately, but :+1: to both the topic and hacking session.

**Comment by frenchy64** [comment]
We found and fixed 3 bugs in about 45 minutes using the power of property-based testing. I'll put it here if anyone wants to continue hacking on it (make a PR against the `madclj` branch) and then we can report it all upstream. Maybe we can spin this into a generic "edn parser parity" library.

Written by @frenchy64 @ericnormand @balloneij @dpetranek 

https://github.com/madclj/fast-edn/pull/1/files

Some observations we discussed:
- Clojure's edn parser is a copy-paste of the Clojure parser with the unsupported syntax deleted
- fast-edn should have a `peek` function for this idiom:
    - https://github.com/tonsky/fast-edn/blob/3a4cdbdae294a5386a56fa1eb59271a472456c22/src/fast_edn/EdnParser.java#L400-L401
- fast-edn manually implements buffered readers with arrays, even specializing them for certain readers:
  - https://github.com/tonsky/fast-edn/blob/3a4cdbdae294a5386a56fa1eb59271a472456c22/src/fast_edn/EdnParser.java#L91-L119
- fast-edn has some fascinating optimizations around bit masking, where clojure.edn uses `IFn[]`
    - https://github.com/tonsky/fast-edn/blob/3a4cdbdae294a5386a56fa1eb59271a472456c22/src/fast_edn/EdnParser.java#L1151-L1182
- fast-edn is beautifully implemented it was a breeze to fix the bugs
- we don't know why the existing data.generative tests didn't pick up the bugs we found, especially `(edn/read-string "{0 -}")`
  - https://github.com/tonsky/fast-edn/blob/3a4cdbdae294a5386a56fa1eb59271a472456c22/test/fast_edn/gen_test.clj#L33-L43

![signal-2025-04-09-225623](https://github.com/user-attachments/assets/1841d58b-9a3a-4d83-a0c9-b5bbe1ed7f2c)

**Reply by brunchboy** [reply]
That's awesome, I wish I'd been able to be a part of it!

**Reply by frenchy64** [reply]
It was a lot of fun and we barely scratched the surface.

@madclj/meetup Anyone want to continue same time next week April 16th 7pm-9pm, while it's still fresh? 👍 if yes.
